<!DOCTYPE html>
<html>
<head><meta charset="utf-8" />

<title>Quick_Start_Guide</title><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>

<style type="text/css">
/* Overrides of notebook CSS for static HTML export */
body {
  overflow: visible;
  padding: 8px;
}
div#notebook {
  overflow: visible;
  border-top: none;
}@media print {
  div.cell {
    display: block;
    page-break-inside: avoid;
  } 
  div.output_wrapper { 
    display: block;
    page-break-inside: avoid; 
  }
  div.output { 
    display: block;
    page-break-inside: avoid; 
  }
}

.cell-kernel-selector {
	width:70pt;
    background: none;
    z-index: 1000;
    position: absolute;
    height: 1.7em;
    margin-top: 3pt;
    right: 8pt;
    font-size: 80%;
}

textarea.sos-source {
  line-height: 1.21429em;
  font-size: 14px;
  background: none;
  width: 100%;
  border: none;
  font-family: monospace;
  height: auto;
  padding: 0.4em;
  outline: none;
  position: relative;
  resize: none;
  overflow: hidden;
  vertical-align: text-top;
}
</style>

<!-- Custom stylesheet, it must be in the same directory as the html file -->

<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>
    <!-- End of mathjax configuration --></head>


<!-- Global Site Tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-107286198-1"></script>
<script>
   window.dataLayer = window.dataLayer || [];
   function gtag(){dataLayer.push(arguments)};
   gtag('js', new Date());
   
   gtag('config', 'UA-107286198-1');
</script>


<meta name="viewport" content="width=device-width, initial-scale=1">




<link rel="stylesheet" href="https://code.jquery.com/ui/1.11.4/themes/smoothness/jquery-ui.css">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<link type="text/css" rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/codemirror.css">
  
<link href="https://fonts.googleapis.com/css?family=Source+Sans+Pro" rel="stylesheet" type="text/css">
<link href="https://fonts.googleapis.com/css?family=Droid+Sans:400,700" rel="stylesheet">
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<style>  /* defined here in case the main.css below cannot be loaded */
   .lev1 {margin-left: 80px}
   .lev2 {margin-left: 100px}
   .lev3 {margin-left: 120px}
   .lev4 {margin-left: 140px}
   .lev5 {margin-left: 160px}
   .lev6 {margin-left: 180px}
   .cm-sos-interpolated {
  background-color: #EDD5F3;
}
.cm-sos-sigil {
  background-color: #EDD5F3;
}
</style>
<link rel="stylesheet" type="text/css" href="../../css/jt.css">
<link rel="stylesheet" type="text/css" href="../../css/toc2.css">
<script src="../../js/doc_toc.js"></script>
<script src="../../js/docs.js"></script>
<script>
   MathJax.Hub.Config({
       "HTML-CSS": {
           preferredFont: "TeX",
           availableFonts: ["STIX","TeX"],
           styles: {
               scale: 110,
               ".MathJax_Display": {
                   "font-size": "110%",
               }
           }
       }
   });
</script>
<script>
   function filterDataFrame(id) {
       var input = document.getElementById("search_" + id);
       var filter = input.value.toUpperCase();
       var table = document.getElementById("dataframe_" + id);
       var tr = table.getElementsByTagName("tr");
   
       // Loop through all table rows, and hide those who don't match the search query
       for (var i = 1; i < tr.length; i++) {
           for (var j = 0; j < tr[i].cells.length; ++j) {
               var matched = false;
               if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                   tr[i].style.display = "";
                   matched = true
                   break;
               }
               if (!matched)
                   tr[i].style.display = "none";
           } 
       }
   }
   
   function sortDataFrame(id, n, dtype) {
       var table = document.getElementById("dataframe_" + id);
   
       var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
       var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
   
       if (dtype === 'numeric') {
           var fn = function(a, b) { 
               return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
           }
       } else {
           var fn = function(a, b) {
               var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
               return c > 0 ? 1 : (c < 0 ? -1 : 0) }
       }
       var isSorted = function(array, fn) {
           if (array.length < 2)
               return 1;
           var direction = fn(array[0], array[1]); 
           for (var i = 1; i < array.length - 1; ++i) {
               var d = fn(array[i], array[i+1]);
               if (d == 0)
                   continue;
               else if (direction == 0)
                   direction = d;
               else if (direction != d)
                   return 0;
               }
           return direction;
       }
   
       var sorted = isSorted(tr, fn);
   
       if (sorted == 1 || sorted == -1) {
           // if sorted already, reverse it
           for(var i = tr.length - 1; i >= 0; --i)
               tb.appendChild(tr[i]); // append each row in order
       } else {
           tr = tr.sort(fn);
           for(var i = 0; i < tr.length; ++i)
               tb.appendChild(tr[i]); // append each row in order
       }
   }
   
   $( document ).ready(function(){
   
               var cfg={'threshold':4,     // depth of toc (number of levels)
                // 'number_sections': true,  // sections numbering
                'number_sections': false, 
                'toc_cell': false,          // useless here
                'toc_window_display': true, // display the toc window
                "toc_section_display": "block", // display toc contents in the window
                // 'sideBar':true,      
                'sideBar':true,       // sidebar or floating window
                'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
               }
   
               var st={};                  // some variables used in the script
               st.rendering_toc_cell = false;
               st.config_loaded = false;
               st.extension_initialized=false;
               st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
               st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
               st.nbcontainer_width = $('#notebook-container').css('width')
               st.oldTocHeight = undefined
               st.cell_toc = undefined;
               st.toc_index=0;
   
               // fire the main function with these parameters
   
   
   
               table_of_contents(cfg,st);
               var file=tutorialsDict[$("h1:first").attr("id")];
               var path="http://vatlab.github.io/sos-docs"
            // $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()

            var tuts=tutorials;
            var pos=tutorials.indexOf(file);

            for (var a=pos;a>=0;a--){
                  var name=tuts[a]
                  $('<li><a href="'+path+'/doc/tutorials/'+name+'.html">'+name.replace(/_/g," ")
                        + '&nbsp ' + '<i class="fa ' +
                      (a === pos ? 'fa-caret-down' : 'fa-caret-right') + '"></i>' + '</a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+path+'/doc/tutorials/'+file+'.html'+'"]').css("color","#126dce");

            for (var a=pos+1;a<tuts.length;a++){
                  var name=tuts[a]
                  $(".toc #toc-level0").append('<li><a href="'+path+'/doc/tutorials/'+name+'.html">'+name.replace(/_/g," ")
                    + '&nbsp' + '<i class="fa fa-caret-right"></i>' +'</a></li>');
            }

            // for (var a =0;a<tuts.length;a++){
            //       var name =tuts[a];
            //       if ()
            //       $(".toc #toc-level0").append('<li><a href="'+path+'/doc/tutorials/'+name+'.html">'+name.replace("_"," ")+'</a></li>');
            // }

            // $(".toc #toc-level0").append('<li id="indexHome"><a href="'+path+'/index.html"><b>Home<b></a></li>');
            // var home=$("#toc-level0 #indexHome");

            // home.insertBefore("#toc-level0 li:eq(0)");


            // $(".number_sections-btn").hide();
            // $(".toc_cell_sections-btn".hide();   
   
       });
</script><style>  /* defined here in case the main.css below cannot be loaded */
   .lan_Bash .input_prompt { background-color: #E6EEFF !important }.lan_R .input_prompt { background-color: #DCDCDA !important }.lan_SoS {}.lan_Julia .input_prompt { background-color: #ebd8eb !important }.lan_MATLAB .input_prompt { background-color: #dff8fb !important }.lan_Python3 .input_prompt { background-color: #FFE771 !important }
</style>


<body>
  <div tabindex="-1" id="notebook" class="border-box-sizing">
    <div class="container" id="notebook-container">

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Quick-Start-Guide">Quick Start Guide<a class="anchor-link" href="#Quick-Start-Guide">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Before we start our tutorial, it is important to understand a SoS script can be executed from command line in batch mode, or interactively in Jupyter notebook. This tutorial is written in Jupyter but you can execute the scripts in batch mode if you save them to disk and execute them using command <code>sos run</code> or <code>sos-runner</code>.</p>
<p>The following commands create some fake input files and fake commands in order to demonstrate the commands. Please refer to <a href="../documentation/Notebook_Interface.html">Chapter Notebook Interface</a> of the SoS documentation if you would like to learn more about them.</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div><div class="inner_cell" style="min-height: 8.44285715em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="5" class="sos-source" name="SoS">%set -v2 -s default
%preview --off
!touch mutated.fastq control.fastq ctrl.fastq case.fastq
!cp STAR Rscript ~/.sos/bin/
!sos remove --signature -v 0</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>Set sos options to &#34;-v2 -s default&#34;
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Multi-Language-Notebook">Multi-Language Notebook<a class="anchor-link" href="#Multi-Language-Notebook">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Let us assume that you are a bioinformaticist needed to compare the expression levels between two samples. You can use a Jupyter notebook with SoS kernel to perform the analysis using different script, the trick here is to select the right kernel for each cell. For example, you can run the following cell in bash if you choose <code>bash</code> as the kernel.</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_Bash">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div><div class="inner_cell" style="min-height: 18.442857160000003em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="12" class="sos-source" name="Bash"># using a fake STAR in ~/
export PATH=$HOME/.sos/bin:$PATH
# index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir STAR_index
# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>Generating STAR_index/chrName.txt
Generating aligned/control.out.tab from control.fastq
Generating aligned/mutated.out.tab from mutated.fastq
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The first command (supposedly) builds an index of the reference genome in preparation for the latter steps, the second command aligns reads from the first sample to the reference genome, and the third command aligns reads from the second sample to the reference genome. Do not panic if you do not know what these commands are doing, this is just an example.</p>
<p>These commands generate, among other files, two files named <code>aligned/control.out.tab</code> and <code>aligned/mutated.out.tab</code> with expression counts of all genes. You then wrote a <a href="https://www.r-project.org/">R</a> script to analyze the results, something like</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_R">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div><div class="inner_cell" style="min-height: 9.871428580000002em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="6" class="sos-source" name="R">control.count <- read.table('aligned/control.out.tab')
mutated.count <- read.table('aligned/mutated.out.tab')
# normalize, compare, output etc, ignored.
pdf('myfigure.pdf')
# plot results
dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">
<strong>pdf:</strong> 2
</div>

</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="String-interpolation">String interpolation<a class="anchor-link" href="#String-interpolation">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>When you work with multiple kernels in the same SoS notebook, the SoS kernel is called the <strong>master kernel</strong> and the rest are called <strong>subkernels</strong>. The SoS kernel is based on a Python syntax so you can define any variables and execute any Python statements such as</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[4]:</div><div class="inner_cell" style="min-height: 4.15714286em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="2" class="sos-source" name="SoS">INDEXDIR = 'STAR_index'
ALIGNDIR = 'aligned'</textarea>
  </div>
</div>

</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Then, a trick to "compose" scripts in subkernels is to use string interpolation to expand expressions inside <code>{ }</code> with their values. For example, with the above definitions, you can execute the commands as</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_Bash">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[5]:</div><div class="inner_cell" style="min-height: 19.87142859em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="13" class="sos-source" name="Bash">%expand
# using a fake STAR in ~/
export PATH=$HOME/.sos/bin:$PATH
# index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir {INDEXDIR}
# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix {ALIGNDIR}/control
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix {ALIGNDIR}/mutated</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>Generating STAR_index/chrName.txt
Generating aligned/control.out.tab from control.fastq
Generating aligned/mutated.out.tab from mutated.fastq
</pre>
</div>
</div>

</div>
</div>

</div>
</div>
<div class="cell border-box-sizing code_cell rendered lan_R">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[6]:</div><div class="inner_cell" style="min-height: 11.300000010000002em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="7" class="sos-source" name="R">%expand
control.count <- read.table('{ALIGNDIR}/control.out.tab')
mutated.count <- read.table('{ALIGNDIR}/mutated.out.tab')
# normalize, compare, output etc, ignored.
pdf('myfigure.pdf')
# plot results
dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">
<strong>pdf:</strong> 2
</div>

</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>using magic <code>%expand</code>. Here we use the default method to expand expressions inside sigil <code>{ }</code>, but if you scripts already have braces, you can use magics such as <code>%expand ${ }</code> to use an alternative sigil.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Data-Exchange-among-subkernels">Data Exchange among subkernels<a class="anchor-link" href="#Data-Exchange-among-subkernels">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>String interpolation is useful for composing scripts to be executed by subkernels, but it would be awkward to pass a large amount of information in this way, and it disallows passing information among subkernels.</p>
<p>SoS provides a few <code>magics</code> to facilitate data exchange among subkernels. For example, the <code>%get</code> magic retrieves information from the SoS kernel to a subkernel, so your shell script could be written as:</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_Bash">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[7]:</div><div class="inner_cell" style="min-height: 17.01428573em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="11" class="sos-source" name="Bash">%get INDEXDIR ALIGNDIR
# index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir $INDEXDIR
# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix $ALIGNDIR/control
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix $ALIGNDIR/mutated</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>Generating STAR_index/chrName.txt
Generating aligned/control.out.tab from control.fastq
Generating aligned/mutated.out.tab from mutated.fastq
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>after retriving <code>INDEXDIR</code> and <code>ALIGNDIR</code> from the SoS kernel and create two native shell variables <code>INDEXDIR</code> and <code>ALIGNDIR</code>. Similarly, your R script could be written as</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_R">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[8]:</div><div class="inner_cell" style="min-height: 11.300000010000002em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="7" class="sos-source" name="R">%get ALIGNDIR
control.count <- read.table(paste0(ALIGNDIR, '/control.out.tab'))
mutated.count <- read.table(paste0(ALIGNDIR, '/mutated.out.tab'))
# normalize, compare, output etc, ignored.
pdf('myfigure.pdf')
# plot results
dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">
<strong>pdf:</strong> 2
</div>

</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The subkernels are persistent so the passed variables can be used in later cells using the same kernel, for example, variable <code>ALIGNDIR</code> can be used in the Bash kernel in new commands as follows:</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_Bash">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[9]:</div><div class="inner_cell" style="min-height: 2.72857143em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="1" class="sos-source" name="Bash">echo Align directory is $ALIGNDIR</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>Align directory is aligned
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Note that SoS magics and multiple live subkernels are only supported by Jupyter notebook and cannot be used in SoS batch mode.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Your-first-SoS-script">Your first SoS script<a class="anchor-link" href="#Your-first-SoS-script">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The project completed successfully and you needed to archive the scripts for later reference. The Jupyter notebook is a perfect format to record both the commands and the results. However, if you would like to only save the steps of your analysis, you can save the notebook in <code>.sos</code> format (<code>File</code>-&gt;<code>Download As</code>-&gt;<code>.sos</code>) or copy/paste the commands to a single SoS script named <code>myanalysis.sos</code>, with content</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[10]:</div><div class="inner_cell" style="min-height: 41.30000004em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="28" class="sos-source" name="SoS">%run
#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

run:
# index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir STAR_index

# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated

R:
control.count <- read.table('aligned/control.out.tab')
mutated.count <- read.table('aligned/mutated.out.tab')
# normalize, compare, output etc, ignored.
pdf('myfigure.pdf')
# plot results
dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre># index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir STAR_index
Generating STAR_index/chrName.txt

# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from control.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from mutated.fastq

Generating myfigure.pdf
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Here <strong><code>run</code></strong> and <strong><code>R</code></strong> are SoS actions that executes the following scripts in <code>bash</code> and <code>R</code> respectively. The scripts are included verbatim and end after reaching another SoS action or directive. For the sake of time, we created fake <code>STAR</code> and <code>R</code> commands that simply states the file generated.</p>
<p>The script is executed in Jupyter when the notebook (this documentation) is executed. In command line, you would need to run the script using command</p>

<pre><code>% sos run myscript</code></pre>
<p>or</p>

<pre><code>% myscript.sos</code></pre>
<p>if you make <code>myscript.sos</code> executable (by running <code>chmod +x myscript.sos</code>.</p>
<p>As you can see, this sos script simply executes the embedded shell and R script and generates the four output files.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Although it is convenient to embed scripts verbatim, it is often clearer to indent the script and write your SoS script as:</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[11]:</div><div class="inner_cell" style="min-height: 42.72857147em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="29" class="sos-source" name="SoS">%run
#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

run:
    # index reference genome
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq\
        --genomeDir STAR_index

    # align reads to the reference genome
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn control.fastq --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn mutated.fastq --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated

# compare expression values
R:
    control.count <- read.table('aligned/control.out.tab')
    mutated.count <- read.table('aligned/mutated.out.tab')
    # normalize, compare, output etc, ignored.
    pdf('myfigure.pdf')
    # plot results
    dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre># index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fastq\
    --genomeDir STAR_index
Generating STAR_index/chrName.txt

# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from control.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from mutated.fastq

Generating myfigure.pdf
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The scripts in this case end with the end of indentation so that you can add some comments for the <code>R</code> script without being considered as part of the previous script. The script works identically to the one without indentation.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Separate-scripts-into-steps">Separate scripts into steps<a class="anchor-link" href="#Separate-scripts-into-steps">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Because the scripts perform different tasks, it is logically clearer to separate them into different <strong>steps</strong>. Actually, because the first reference-generating command is not a data processing step, it makes sense to separate it into two scripts. Now we can insert step headers to the script as follows:</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[12]:</div><div class="inner_cell" style="min-height: 48.44285719em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="33" class="sos-source" name="SoS">%run
#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

[1]
# index reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index

[2]
# align reads to the reference genome
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn control.fastq --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn mutated.fastq --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated

[3]
# compare expression values
R:
    control.count <- read.table('aligned/control.out.tab')
    mutated.count <- read.table('aligned/mutated.out.tab')
    # normalize, compare, output etc, ignored.
    pdf('myfigure.pdf')
    # plot results
    dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir STAR_index
Generating STAR_index/chrName.txt

STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from control.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from mutated.fastq

Generating myfigure.pdf
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now, when you execute the script with command</p>

<pre><code>sos run myscript</code></pre>
<p>SoS will display <code>default_1</code>, <code>default_2</code> and <code>default_3</code> to report progress. The comments after section heads are considered step comments and will also be displayed during execution.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>It is worth noting that you can write SoS workflows in Jupyter notebook as long as you specify a header for each step. For example, you can create a Jupyter notebook with the following cells</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[13]:</div><div class="inner_cell" style="min-height: 8.44285715em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="5" class="sos-source" name="SoS">[1]
# index reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index</textarea>
  </div>
</div>

</div>

</div>
</div>
<div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[14]:</div><div class="inner_cell" style="min-height: 14.157142870000001em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="9" class="sos-source" name="SoS">[2]
# align reads to the reference genome
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn control.fastq --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn mutated.fastq --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated</textarea>
  </div>
</div>

</div>

</div>
</div>
<div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[15]:</div><div class="inner_cell" style="min-height: 14.157142870000001em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="9" class="sos-source" name="SoS">[3]
# compare expression values
R:
    control.count <- read.table('aligned/control.out.tab')
    mutated.count <- read.table('aligned/mutated.out.tab')
    # normalize, compare, output etc, ignored.
    pdf('myfigure.pdf')
    # plot results
    dev.off()</textarea>
  </div>
</div>

</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>and execute the workflow with command</p>

<pre><code>sos run myscript.ipynb</code></pre>
<p>The notebook can contain markdown cells and cells with other kernels, and sos cells without section header, and the <code>sos run</code> (and other) commands will ignore all such cells and only extract and execute workflows defined in the notebook.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Make-the-script-work-for-other-input-files">Make the script work for other input files<a class="anchor-link" href="#Make-the-script-work-for-other-input-files">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>After a while, before you almost forgot about this analysis, you needed to analyze another pair of samples. You could copy <code>myanalysis.sos</code> to <code>myanalysis2.sos</code>, change filenames and run it, but an easier way is to change your SoS file to accommodate other input files. This can be done by defining a command line argument and passing files name to a <strong>SoS variable</strong>:</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[16]:</div><div class="inner_cell" style="min-height: 54.15714291em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="37" class="sos-source" name="SoS">%run
#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fastq formats. The first one for control sample
# and the second one for mutated sample.
parameter: fastq_files=['control.fastq', 'mutated.fastq']

[1]
# index reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index

[2]
# align reads to the reference genome
run: expand=True
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {fastq_files[0]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {fastq_files[1]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated

[3]
# compare expression values
R: expand=True
    control.count <- read.table('aligned/control.out.tab')
    mutated.count <- read.table('aligned/mutated.out.tab')
    # normalize, compare, output etc, ignored.
    pdf('myfigure.pdf')
    # plot results
    dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir STAR_index
Generating STAR_index/chrName.txt

STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from control.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from mutated.fastq

Generating myfigure.pdf
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>A command line argument <code>fastq_files</code> is defined in with <code>parameter</code> keyword. With this definition, you can pass two filenames to variable <code>fastq_files</code> from command line</p>
<div class="highlight"><pre><span></span>sos run myanalysis --fastq_files ctrl.fastq <span class="k">case</span>.fastq
</pre></div>
<p><code>{fastq_files[0]}</code> and <code>{fastq_files[1]}</code> in command <code>STAR --genomeDir ...</code> will be replaced with their values before the commands are executed. Here <code>fastq_files[0]</code> and <code>fastq_files[1]</code> are Python expressions that will be evaluated during execution.</p>
<p>In Jupyter notebook, we can execute the previous cell with additional option using magic <code>%rerun</code>.</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[17]:</div><div class="inner_cell" style="min-height: 2.72857143em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="1" class="sos-source" name="SoS">%rerun --fastq_files ctrl.fastq case.fastq</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
    --genomeDir STAR_index
Generating STAR_index/chrName.txt

STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn ctrl.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from ctrl.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn case.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from case.fastq

Generating myfigure.pdf
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Note that the <code>parameter</code> statement is defined before any SOS step, actually in a special <code>[global]</code> section although the section is usually ignored in <code>.sos</code> format.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Ignore-steps-that-do-not-need-to-be-rerun">Ignore steps that do not need to be rerun<a class="anchor-link" href="#Ignore-steps-that-do-not-need-to-be-rerun">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Although the SoS script now accepts command line arguments, it is still no more than a compilation of scripts and you immediately realized that it is a waste of time to execute the first command each time. To solve this problem, you can convert the SoS script to a real workflow by telling SoS the input and output of each step:</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[18]:</div><div class="inner_cell" style="min-height: 62.72857149em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="43" class="sos-source" name="SoS">%run -v4
#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fastq formats. The first one for control sample
# and the second one for mutated sample.
parameter: fastq_files=['control.fastq', 'mutated.fastq']

[1]
# create a index for reference genome
output: 'STAR_index/chrName.txt'
run: 
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index

[2]
# align the reads to the reference genome
input:    fastq_files
depends:  'STAR_index/chrName.txt'
output:   ['aligned/control.out.tab', 'aligned/mutated.out.tab']
run:  expand=True
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input[0]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input[1]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated

[3]
# compare expression values
print(input)
output: 'myfigure.pdf'
R:  expand=True
    control.count <- read.table('{input[0]}')
    mutated.count <- read.table('{input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('{_output}')
    # plot results
    dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>INFO: <span class="ansi-green-fg">default_1</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
</pre>
</div>
</div>

<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from control.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from mutated.fastq

control.fastq mutated.fastq
</pre>
</div>
</div>

<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>INFO: <span class="ansi-green-fg">default_3</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Here we</p>
<ul>
<li>Use <strong>output directive</strong> to specify the expected output of all steps.</li>
<li>Use <strong>input directive</strong> to specify the input of step 2. Step 1 by default has no input and input for step 3 by default is the output of step 2, its previous step.</li>
<li>Use <strong>depends directive</strong> to let step 2 depend on the output of step 1.</li>
<li>Use <code>{_input[0]}</code> and <code>{_input[1]}</code> in step 2 and 3 because these steps now have properly-defined <code>_input</code>. This variable is defined by step input as the input file of the step.</li>
</ul>
<p>With such information, when you run the same command with another set of input file</p>
<div class="highlight"><pre><span></span>sos run myanalysis --input ctrl.fastq <span class="k">case</span>.fastq
</pre></div>
<p>SoS will ignore step 1 if this step has been run with output <code>STAR_index/chrName.txt</code>. The same happens to step 2 and 3 so all steps will be ignored if you run the script repeatedly with the same input and processing scripts. SoS uses <strong>runtime signature</strong> for each step and will re-run the step if and only if the content or filename of input, output files or the processing scripts are changed.</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[19]:</div><div class="inner_cell" style="min-height: 2.72857143em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="1" class="sos-source" name="SoS">%rerun --fastq-files ctrl.fastq case.fastq</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>INFO: <span class="ansi-green-fg">default_1</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
</pre>
</div>
</div>

<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn ctrl.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from ctrl.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn case.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from case.fastq

ctrl.fastq case.fastq
Generating myfigure.pdf
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>An added benefit of specifying input and output of steps is it allows you to create an archive of your project with all input, output, and intermediate files. This can be achived using command</p>
<div class="highlight"><pre><span></span>% sos pack -o myanalysis.sar
</pre></div>
<p>if only one workflow has been executed under the current directory. The output is a compressed archive with extension <code>.sar</code> and can be examined and unpacked using command <code>sos unpack</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Use-make-rule-to-define-resource-providing-steps">Use make-rule to define resource-providing steps<a class="anchor-link" href="#Use-make-rule-to-define-resource-providing-steps">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Instead of using runtime signature to avoid re-running the first step, we can also make step 1 an optional step that will be executed only necessary. That is to say, we can define this step as an <code>auxiliary step</code> that will only be called when the file it <strong>provides</strong> does not exist.</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[20]:</div><div class="inner_cell" style="min-height: 61.30000006em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="42" class="sos-source" name="SoS">%run

#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fastq formats. The first one for control sample
# and the second one for mutated sample.
parameter: fastq_files=['control.fastq', 'mutated.fastq']

[build_index: provides='STAR_index/chrName.txt']
# create a index for reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index

[1]
# align the reads to the reference genome
input:    fastq_files
depends:  'STAR_index/chrName.txt'
output:   ['aligned/control.out.tab', 'aligned/mutated.out.tab']
run:  expand=True
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input[0]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input[1]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated

[2]
# compare expression values
output: 'myfigure.pdf'
R:  expand=True
    control.count <- read.table('{input[0]}')
    mutated.count <- read.table('{input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('{_output}')
    # plot results
    dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn control.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/control
Generating aligned/control.out.tab from control.fastq
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
    --readFilesIn mutated.fastq --quantMode GeneCounts \
    --outFileNamePrefix aligned/mutated
Generating aligned/mutated.out.tab from mutated.fastq

</pre>
</div>
</div>

<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>INFO: <span class="ansi-green-fg">default_2</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The new script consists of two steps, and an auxiliary step <code>build_index</code> that will only be executed when <code>STAR_index/chrName.txt</code> is not available. A slight difference between this version and the previous one is that while the previous version will always execute the first step, this version will not execute it if <code>STAR_index/chrName.txt</code> has been generated before in any way, perhaps not by SoS.</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[21]:</div><div class="inner_cell" style="min-height: 2.72857143em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="1" class="sos-source" name="SoS">%rerun</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>INFO: <span class="ansi-green-fg">default_1</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
INFO: <span class="ansi-green-fg">default_2</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Execute-long-running-jobs-externally">Execute long-running jobs externally<a class="anchor-link" href="#Execute-long-running-jobs-externally">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The first step will takes a long time to execute. Instead of executing them by SoS, you might want to submit the commands to a job-queue and be executed and monitored externally. This is especially useful when you need to execute multiple SoS workflows on a cluster-based system. Without going through the details on how to set up your job-queue (see another tutorial for details), it is almost trivial to modify your script to define part of a <strong>step process</strong> to a <strong>task</strong>:</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[22]:</div><div class="inner_cell" style="min-height: 62.72857149em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="43" class="sos-source" name="SoS">%run

#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fastq formats. The first one for control sample
# and the second one for mutated sample.
parameter: fastq_files=['control.fastq', 'mutated.fastq']

[build_index: provides='STAR_index/chrName.txt']
# create a index for reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index

[1]
# align the reads to the reference genome
input:    fastq_files
depends:  'STAR_index/chrName.txt'
output:   ['aligned/control.out.tab', 'aligned/mutated.out.tab']
task:
run:  expand=True
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input[0]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input[1]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated

[2]
# compare expression values
output: 'myfigure.pdf'
R:  expand=True
    control.count <- read.table('{input[0]}')
    mutated.count <- read.table('{input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('{_output}')
    # plot results
    dev.off()</textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>INFO: <span class="ansi-green-fg">default_1</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
INFO: <span class="ansi-green-fg">default_2</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>As you can see, the only difference is the insertion of <code>task:</code> directive before <code>run</code>. Now, when you execute the command, the <code>STAR</code> commands will be executed externally and you can monitor the status of the jobs using your web browser. If no job-queue is set up, the command will be executed in a separate process.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Execute-steps-in-parallel">Execute steps in parallel<a class="anchor-link" href="#Execute-steps-in-parallel">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Although the two steps of this example have to be executed sequentially, the first step runs the <code>STAR</code> command twice on two input files, and can be executed in parallel. You can tell this to SoS by modifying the script as follows</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[23]:</div><div class="inner_cell" style="min-height: 58.4428572em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="40" class="sos-source" name="SoS">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fastq formats. The first one for control sample
# and the second one for mutated sample.
parameter: fastq_files=['control.fastq', 'mutated.fastq']

[build_index: provides='STAR_index/chrName.txt']
# create a index for reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index

[1]
# align the reads to the reference genome
sample_type = ['control', 'mutated']
input:    fastq_files, group_by='single', paired_with='sample_type'
depends:  'STAR_index/chrName.txt'
output:   f"aligned/{_sample_type}.out.tab"

task:     concurrent=True
run:  expand=True
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input:q} --quantMode GeneCounts \
        --outFileNamePrefix aligned/{_sample_type}

[2]
# compare expression values
output: 'myfigure.pdf'
R:  expand=True
    control.count <- read.table('{input[0]}')
    mutated.count <- read.table('{input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('{_output}')
    # plot results
    dev.off()
</textarea>
  </div>
</div>

</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Here we</p>
<ol>
<li>Use option <code>group_by='single'</code> to pass input one by one to action. The action will be executed twice with <code>_input</code> set to the first and second input file respectively.</li>
<li>Define a variable <code>sample_type</code> and pair it with input files (option <code>paired_with</code>). This will generate a variable <code>_sample_type</code> for each input file so <code>_sample_type</code> will be <code>control</code> for the first input file, and <code>mutated</code> for the second.</li>
<li>Use <code>{_input}</code> and <code>{_sample_type}</code> to define partial <code>output</code>.</li>
<li>Use <code>{_input:q}</code> instead of <code>{_input}</code> in the script. This is a small trick to shell-quote filenames so that filenames with spaces and other special characters can be properly quoted in shell commands.</li>
</ol>
<p>Now, if you execute the script with option <code>-j 2</code> (2 concurrent processes),</p>
<div class="highlight"><pre><span></span>sos run myanalysis.sos --input control1.fastq control2.fastq -j <span class="m">2</span>
</pre></div>
<p>the second step submit two jobs to job-queue, or execute them in two separate processes if a job-queue is not setup.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Generating-a-report">Generating a report<a class="anchor-link" href="#Generating-a-report">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In addition to functions such as <code>run</code> and <code>R</code> that execute scripts in different languages, SoS provides a number of other functions (called actions) that can be used to, for example, generate reports using command <code>pandoc</code> or <code>Rmarkdown</code>. For example, by adding a step with a <code>pandoc</code> action, you can compose a script in Markdown format and generate a nice-looking report in HTML format.</p>

</div>
</div>
</div><div class="cell border-box-sizing code_cell rendered lan_SoS">
   
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[24]:</div><div class="inner_cell" style="min-height: 82.72857151000001em;">
  <div class="input_area cm-s-ipython">
   <textarea rows="57" class="sos-source" name="SoS">%preview report.html
%run

#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fastq formats. The first one for control sample
# and the second one for mutated sample.
parameter: fastq_files=['control.fastq', 'mutated.fastq']

[build_index: provides='STAR_index/chrName.txt']
# create a index for reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fastq \
        --genomeDir STAR_index

[1]
# align the reads to the reference genome
sample_type = ['control', 'mutated']
input:    fastq_files, group_by='single', paired_with='sample_type'
depends:  'STAR_index/chrName.txt'
output:   f"aligned/{_sample_type[0]}.out.tab"

run:  expand=True
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn {_input:q} --quantMode GeneCounts \
        --outFileNamePrefix aligned/{_sample_type[0]}

report:   output='align.md', active=-1,  expand=True
    ## Alignment of reads
    * input file: {_input}
    * output file: {_output}

[2]
# compare expression values
output: 'myfigure.pdf'
R:  expand=True
    control.count <- read.table('{_input[0]}')
    mutated.count <- read.table('{_input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('{_output}')
    # plot results
    dev.off()

report:  output='analysis.md', active=-1,  expand=True
    ## Statistical analysis
    * input file: {_input}
    * Figure: {_output}

[3]
pandoc: input=['align.md', 'analysis.md'], output='report.html'
    # An analysis of two RNA Seq samples
    </textarea>
  </div>
</div>

</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

<div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>INFO: <span class="ansi-green-fg">default_1</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
INFO: <span class="ansi-green-fg">default_1</span> (index=1) is <span class="ansi-green-fg">ignored</span> due to saved signature
INFO: <span class="ansi-green-fg">default_2</span> (index=0) is <span class="ansi-green-fg">ignored</span> due to saved signature
INFO: Report saved to report.html
</pre>
</div>
</div>

</div>
</div>

</div>
</div>

<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In this example, partial reports are prepared whenever a step completes, and a final step is used combine all partial reports (with a header) and process it with pandoc.</p>
<p>Please refer to <a href="Generating_Reports.html">the report generation tutorial</a> for details.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We have showed you multiple versions of the same SoS script, each using more features of SoS. This actually demonstrates one of the advantages of the SoS system, namely you can start using SoS in minutes without knowing any of its advanced features, and gradually improve your script if needed.</p>

</div>
</div>
</div>
    </div>
  </div>
</body>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/codemirror.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/python/python.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/r/r.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/octave/octave.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/ruby/ruby.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/sas/sas.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/javascript/javascript.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/shell/shell.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/julia/julia.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.38.0/mode/markdown/markdown.js"></script>
	  <script src="../../js/sos-mode.js"> </script>
      <script>
		   function highlight_cells(cells, i, interval) {
			  setTimeout(function() {
				var editor = CodeMirror.fromTextArea(cells[i], {
		           lineNumbers: false,
				   styleActiveLine: true,
		           matchBrackets: true,
				   readOnly: true,
		           mode: 'sos',
				   base_mode: cells[i].name,
		         });
				$(cells[i]).parent().prepend("<select class='cell-kernel-selector'><option>" 
					+ cells[i].name + "</option></select>");
		      if (i < cells.length)
			    highlight_cells(cells, i + 1, interval);
			}, interval);
		  }

	      highlight_cells(document.getElementsByClassName("sos-source"), 0, 10);
      </script>



 


</html>
